### PR TITLE
Incrementally, ROAs, CAC, etc...

### LINKED ISSUE #2211: Compute ROAs the correct way (like in the google paper)
Our current implementation of ROAs is very simplistic:
```
ROAs[t] = contribution[t] / spend[t]
```
But the ROAS computation has to follow formula (10) in https://storage.googleapis.com/gweb-research2023-media/pubtools/3806.pdf . In particular, note the summation indices where we need to consider l_max from the adstock effect. 
We would like to implement it in the correct way

### PR DESCRIPTION
**Pull request by isofer**
## Description

The research for this feature has been split into two PRs to reduce complexity. This first PR focuses on the core `Incrementality` module and its tests; a follow-up PR will handle shared infrastructure extraction, summary integration, and documentation.

Adds a new `Incrementality` module (`pymc_marketing/mmm/incrementality.py`) that computes **incremental channel contributions** using counterfactual analysis, properly accounting for adstock carryover effects.

The existing ROAS computation (`MMMIDataWrapper.get_roas()`) uses naïve element-wise division (contribution / spend), which ignores the temporal attribution from adstock transformations — spend at time *t* affects outcomes at *t, t+1, ..., t+l_max*. This PR implements the counterfactual approach from the Google MMM paper (Section 4.1), which compares the model's prediction under actual spend vs. a counterfactual where channel spend is removed or perturbed, summing over the full carryover window.

### Technical approach

The `Incrementality` class uses `extract_response_distribution()` + `vectorize_graph()` to compile a single batched evaluator that processes all counterfactual scenarios at once. For each period defined by the requested frequency:

1. Build a zero-padded counterfactual window covering `[t0 - l_max, t1 + l_max]` to capture carry-in history and carry-out effects
2. Apply the `counterfactual_spend_factor` only to spend within `[t0, t1]`, keeping surrounding spend intact for correct adstock context
3. Evaluate baseline on the full dataset once, then batch-evaluate all counterfactual windows
4. Sum the contribution difference over each period's evaluation window (including carryover when enabled)

The module exposes four public methods:
- `compute_incremental_contribution()` — core function, supports total (factor=0) and marginal (factor=1.01) incrementality
- `contribution_over_spend()` — ROAS (or customers/dollar, etc.)
- `spend_over_contribution()` — CAC (reciprocal of above)
- `marginal_contribution_over_spend()` — marginal ROAS at current operating point

> [!NOTE]
> **Key review focus**: The most important thing to review is `test_compute_incremental_contribution_matches_ground_truth` and its helper functions (`evaluate_channel_contribution` and `compute_ground_truth_incremental_by_period`). These validate the vectorized implementation against a completely independent oracle that uses `sample_posterior_predictive` — if the ground truth logic is correct, the rest follows. see more details below. 

## Related Issue

- Getting us closer to close #2211

## Differences from Research Architecture

The implementation is based on the research document [`channel-attribution-architecture.md`](thoughts/shared/research/2026-02-06_14-21-06_channel-attribution-architecture.md). Key differences from the proposed architecture:

| Aspect | Change Type | Research Proposal | Implementation |
|--------|-------------|------------------|----------------|
| **Carryover flags** | Simplification | Separate `include_carryin` and `include_carryout` parameters | Single `include_carryover` boolean. For incrementality, a period either affects adjacent periods or it doesn't — separate carry-in/carry-out flags make sense for saturation curves but not here. |
| **Baseline evaluation** | Optimization | Research pseudocode re-evaluates baseline per period via batched `[baseline, counterfactual]` pairs | Implementation evaluates baseline once on the full dataset (more efficient), then evaluates only counterfactual scenarios in the batch. |
| **Zero-padded windowing** | Research bug fix | Not explicitly described | Implementation zero-pads counterfactual windows at dataset boundaries and right-pads to uniform size for batched evaluation. This handles edge cases where the carryover window extends beyond the dataset. |
| **`response_variable`** | Research bug fix | Research used `"y"` (full model prediction) | Implementation uses `"channel_contribution"` (channel-level contributions), which is the correct level for per-channel incrementality. |

## What Is Complete vs. What Follows

To reduce the complexity of this PR, some related work has been split out into a follow-up PR.

### Complete in this PR

- **`Incrementality` class** with all four public methods (`compute_incremental_contribution`, `contribution_over_spend`, `spend_over_contribution`, `marginal_contribution_over_spend`)
- **`.incrementality` property** on `MMM` class following the established facade pattern (`.plot`, `.summary`, `.data`, `.sensitivity`)
- **Comprehensive test suite** (24 test cases) with ground truth validation against an independent `sample_posterior_predictive` oracle, covering multiple frequencies, panel models, total/marginal incrementality, and edge cases

### Follow-up PR

- **Shared infrastructure + SensitivityAnalysis refactor**: Extract reusable vectorization utilities to a shared module (`counterfactual_core.py`) and refactor `SensitivityAnalysis` to use them, removing duplicated dimension management code
- **Summary integration**: Update `summary.roas()` to delegate to `Incrementality.contribution_over_spend()` for carryover-corrected ROAS
- **Deprecation**: Add deprecation warning to `MMMIDataWrapper.get_roas()` pointing users to `mmm.incrementality.contribution_over_spend()`
- **Documentation**: API reference docs, example notebook demonstrating ROAS/CAC/mROAS workflows

### Incrementality (core of this PR)

- **New**: `pymc_marketing/mmm/incrementality.py` — `Incrementality` class with counterfactual-based incremental contribution, ROAS, CAC, and marginal ROAS
- **New**: `tests/mmm/test_incrementality.py` — comprehensive test suite with ground truth oracle validation
- **Modified**: `pymc_marketing/mmm/multidimensional.py` — added `.incrementality` property
- **Modified**: `tests/mmm/conftest.py` — updated fixtures to use weekly frequency and proper `fit_data` construction via `create_fit_data()`

### Unrelated minor change

- **Modified**: `pymc_marketing/data/idata/mmm_wrapper.py` — added `from_mmm()` classmethod to reduce schema-building duplication, minor type annotation updates
- **Modified**: `pymc_marketing/mmm/multidimensional.py` — refactored `.data` property to use `MMMIDataWrapper.from_mmm()`

## How to Test

1. Run the incrementality test suite:
   ```bash
   pytest tests/mmm/test_incrementality.py -v
   ```

2. Verify existing tests still pass:
   ```bash
   pytest tests/mmm/ -v --ignore=tests/mmm/test_incrementality.py
   ```

3. Manual verification with a fitted model:
   ```python
   # After fitting an MMM
   incr = mmm.incrementality

   # Total incremental contributions
   ic = incr.compute_incremental_contribution(frequency="quarterly")

   # ROAS
   roas = incr.contribution_over_spend(frequency="quarterly")

   # Marginal ROAS
   mroas = incr.marginal_contribution_over_spend(frequency="quarterly")

   # Verify marginal < total for saturating channels
   assert (mroas.mean("sample") < roas.mean("sample")).any()
   ```

### CODE REVIEW

File: pymc_marketing/mmm/incrementality.py [hunk 1]
```diff
@@ -0,0 +1,883 @@
...
+        }
+
+        dates = pd.date_range(start, end, freq="D")
+        periods = dates.to_period(freq_map[frequency])
+        unique_periods = periods.unique()
+
+        period_ranges = []
+        for period in unique_periods:
+            period_start = period.to_timestamp()
+            period_end = period.to_timestamp(how="end")  # Period-end date
```
[the remark this answers is not in the document]
Comment (isofer) [reply pymc_marketing/mmm/incrementality.py]: fixed

File: pymc_marketing/mmm/incrementality.py [hunk 2]
```diff
@@ -0,0 +1,883 @@
...
+        if random_state is None:
+            rng = np.random.default_rng()
+        elif isinstance(random_state, (int, np.integer)):
+            rng = np.random.default_rng(random_state)
+        else:
+            rng = random_state
+
+        # Sample without replacement
+        indices = rng.choice(total_draws, size=num_samples, replace=False)
+        return np.sort(indices)
```
[the remark this answers is not in the document]
Comment (isofer) [reply pymc_marketing/mmm/incrementality.py]: fixed

File: pymc_marketing/mmm/incrementality.py [hunk 3]
```diff
@@ -0,0 +1,901 @@
...
+        period_ranges = []
+        for period in unique_periods:
+            period_start = period.to_timestamp()
+            period_end = period.to_timestamp(how="end").normalize()
+
+            # Clip start to the requested range (needed when the user
+            # passes a start date inside a period)
+            period_start = max(period_start, start)
+
+            period_ranges.append((period_start, period_end))
```
[the remark this answers is not in the document]
Comment (isofer) [reply pymc_marketing/mmm/incrementality.py:1158]: narrow edge case. fixed

File: pymc_marketing/mmm/incrementality.py [hunk 4]
```diff
@@ -0,0 +1,916 @@
...
+
+        period_start_ts: pd.Timestamp = (
+            dates[0] if period_start is None else pd.to_datetime(period_start)
+        )
+        period_end_ts: pd.Timestamp = (
+            dates[-1] if period_end is None else pd.to_datetime(period_end)
+        )
+
+        # Create period groups based on frequency
+        periods = self._create_period_groups(period_start_ts, period_end_ts, frequency)
```
[the remark this answers is not in the document]
Comment (isofer) [reply pymc_marketing/mmm/incrementality.py:338]: fixed

File: pymc_marketing/mmm/incrementality.py [hunk 5]
```diff
@@ -0,0 +1,916 @@
...
+            "yearly": "Y",
+        }
+
+        dates = pd.date_range(start, end, freq="D")
+        periods = dates.to_period(freq_map[frequency])
+        unique_periods = periods.unique()
+
+        # Validate that end aligns with a period boundary.
+        last_period_boundary = unique_periods[-1].to_timestamp(how="end").normalize()
+        if end < last_period_boundary:
```
[the remark this answers is not in the document]
Comment (isofer) [reply pymc_marketing/mmm/incrementality.py:1136]: fixed

File: pymc_marketing/mmm/incrementality.py [hunk 6]
```diff
@@ -0,0 +1,935 @@
...
+        self.idata = idata
+        self.data = MMMIDataWrapper.from_mmm(model, idata)
+
+    def compute_incremental_contribution(
+        self,
+        frequency: Frequency,
+        period_start: str | pd.Timestamp | None = None,
+        period_end: str | pd.Timestamp | None = None,
+        include_carryover: bool = True,
+        original_scale: bool = True,
```
Comment (juanitorduz) [inline pymc_marketing/mmm/incrementality.py]: IMO we always want all these incrementality metrics in the original scale. Shall we remove this parameter and always work in the original scale (can be easier to maintain).

File: pymc_marketing/mmm/incrementality.py [hunk 7]
```diff
@@ -0,0 +1,935 @@
...
+        # Subsample posterior if needed (correctly across chain x draw)
+        idata_sub = self._get_subsampled_idata(num_samples, random_state)
+
+        # Extract response distribution (batched over samples)
+        response_graph = extract_response_distribution(
+            pymc_model=self.model.model,
+            idata=idata_sub,
+            response_variable="channel_contribution",
+        )
+        # Shape: (sample, date, channel, *custom_dims)
```
Comment (juanitorduz) [inline pymc_marketing/mmm/incrementality.py]: for integraton with arviz it might be better to return `(chain, draw, date, channel, *custom_dims)`, would it be possible to work with these dims?

File: pymc_marketing/mmm/incrementality.py [hunk 8]
```diff
@@ -0,0 +1,935 @@
...
+        if period_end_ts > data_end:
+            raise ValueError(
+                f"period_end '{period_end_ts.date()}' is after fitted data "
+                f"end '{data_end.date()}'."
+            )
+        if period_start_ts > period_end_ts:
+            raise ValueError(
+                f"period_start '{period_start_ts.date()}' is after "
+                f"period_end '{period_end_ts.date()}'."
+            )
```
Comment (juanitorduz) [inline pymc_marketing/mmm/incrementality.py]: shall we move this to a simpler function `self._validate_input_dates(...)` ?

File: pymc_marketing/mmm/incrementality.py [hunk 9]
```diff
@@ -0,0 +1,935 @@
...
+            # (represents "no spend" before the dataset)
+            left_pad = 0
+            if n_actual > 0 and actual_dates[0] > ideal_start:
+                left_pad = round((actual_dates[0] - ideal_start) / freq_td)
+
+            # Right-padding: ideal positions after the last actual date
+            # (represents "no spend" after the dataset)
+            right_pad = 0
+            if n_actual > 0 and actual_dates[-1] < ideal_end:
+                right_pad = round((ideal_end - actual_dates[-1]) / freq_td)
```
Comment (juanitorduz) [inline pymc_marketing/mmm/incrementality.py]: Can we make a smaller function with this logi + add unit tests? (especially for edge cases)

File: pymc_marketing/mmm/incrementality.py [hunk 10]
```diff
@@ -0,0 +1,935 @@
...
+                eval_offsets = np.where(eval_in_actual)[0] + start_pos
+                cf_mask[eval_offsets] = True
+            cf_eval_masks.append(cf_mask)
+
+        # Evaluate all counterfactuals at once
+        cf_array = np.stack(cf_scenarios, axis=0)
+        # Shape: (n_periods, max_window, channel, *custom_dims)
+
+        cf_predictions = evaluator(cf_array)
+        # Shape: (n_periods, n_samples, max_window, channel, *custom_dims)
```
Comment (juanitorduz) [inline pymc_marketing/mmm/incrementality.py]: Same here on modularity: extract this code in a function + unit test (it is hard to review within this large function.

File: pymc_marketing/mmm/incrementality.py [hunk 11]
```diff
@@ -0,0 +1,935 @@
...
+            # Single period, no date dimension
+            result = results[0].squeeze("date", drop=True)
+        else:
+            result = xr.concat(results, dim="date")
+
+        # Ensure standard dimension order
+        dim_order = ["sample", "date", "channel", *self.model.dims]
+        if frequency == "all_time":
+            dim_order.remove("date")
+        result = result.transpose(*dim_order)
```
Comment (juanitorduz) [inline pymc_marketing/mmm/incrementality.py]: This inside a helper function + unitest

File: pymc_marketing/mmm/incrementality.py [hunk 12]
```diff
@@ -0,0 +1,935 @@
...
+
+        if num_samples >= total_samples:
+            return self.idata
+
+        rng = np.random.default_rng(random_state)
+        flat_indices = rng.choice(total_samples, size=num_samples, replace=False)
+
+        stacked = posterior.stack(sample=("chain", "draw"))
+        selected = stacked.isel(sample=flat_indices)
+        # Drop MultiIndex components, rename sample -> draw, add chain dim
```
Comment (juanitorduz) [inline pymc_marketing/mmm/incrementality.py]: From a comment above: for arviz integration, I think it might be better to keep (chain, draw)

**Review by juanitorduz — CHANGES_REQUESTED** [review]
This looks super nice! I really like the API @isofer . I left some initial comments on modularization because its a bit hard to review and keep the logic in my head :)

File: pymc_marketing/mmm/incrementality.py (hunk 6 above)
Comment (isofer) [reply pymc_marketing/mmm/incrementality.py]: Agreed — removed the `original_scale` parameter. The module now always works in original scale, which simplifies the API and maintenance.

File: pymc_marketing/mmm/incrementality.py (hunk 7 above)
Comment (isofer) [reply pymc_marketing/mmm/incrementality.py]: Done — restructured the output to return `(chain, draw, date, channel, *custom_dims)`. Internally the vectorized evaluation still uses a flattened sample dimension for efficiency, but at the output stage we reshape back to `(chain, draw)`. When `num_samples` is used (subsampling across the flattened chain x draw space), the result is returned as `(chain=1, draw=num_samples, ...)` since chain provenance is lost during random subsampling.

File: pymc_marketing/mmm/incrementality.py (hunk 8 above)
Comment (isofer) [reply pymc_marketing/mmm/incrementality.py]: Done — extracted to `_validate_input_dates(period_start, period_end) -> tuple[pd.Timestamp, pd.Timestamp]`. Added unit tests covering: None defaults to data bounds, start before data raises, end after data raises, reversed range raises.

File: pymc_marketing/mmm/incrementality.py (hunk 9 above)
Comment (isofer) [reply pymc_marketing/mmm/incrementality.py]: Done — extracted to `_compute_window_metadata(periods, dates, l_max, freq_offset, freq) -> tuple[list[dict], int]`. Added unit tests for interior periods (no padding) and boundary periods (left/right padding).

File: pymc_marketing/mmm/incrementality.py (hunk 10 above)
Comment (isofer) [reply pymc_marketing/mmm/incrementality.py]: Done — extracted to `_build_counterfactual_scenarios(...) -> tuple[np.ndarray, list[np.ndarray], list[pd.Timestamp]]`. Added unit tests for correct output shape and counterfactual factor application. The main method is now a high-level orchestrator: validate -> create periods -> compute windows -> build counterfactuals -> evaluate -> assemble results.

File: pymc_marketing/mmm/incrementality.py (hunk 11 above)
Comment (isofer) [reply pymc_marketing/mmm/incrementality.py]: Done — extracted to `_compute_period_increments(...)` which handles per-period summing, sign convention, xarray construction, and reshaping back to `(chain, draw)`.

File: pymc_marketing/mmm/incrementality.py (hunk 12 above)
Comment (isofer) [reply pymc_marketing/mmm/incrementality.py]: Done — same as the earlier comment, output now uses `(chain, draw)` throughout. See reply above for details on the subsampling case.

File: pymc_marketing/mmm/incrementality.py [hunk 13]
```diff
@@ -0,0 +1,1170 @@
...
+        posterior = self.idata.posterior
+        n_chains = posterior.sizes["chain"]
+        n_draws = posterior.sizes["draw"]
+        total_samples = n_chains * n_draws
+
+        if num_samples >= total_samples:
+            return self.idata
+
+        rng = np.random.default_rng(random_state)
+        flat_indices = rng.choice(total_samples, size=num_samples, replace=False)
```
[the remark this answers is not in the document]
Comment (isofer) [reply pymc_marketing/mmm/incrementality.py]: not an issue in numpy>=2.x

File: pymc_marketing/mmm/incrementality.py [hunk 14]
```diff
@@ -0,0 +1,1170 @@
...
+        self.idata = idata
+        self.data = MMMIDataWrapper.from_mmm(model, idata)
+
+    def compute_incremental_contribution(
+        self,
+        frequency: Frequency,
+        period_start: str | pd.Timestamp | None = None,
+        period_end: str | pd.Timestamp | None = None,
+        include_carryover: bool = True,
+        num_samples: int | None = None,
```
Comment (cetagostini) [inline pymc_marketing/mmm/incrementality.py:194]: Can you change by `posterior_sample_fraction` thats the name in sensitivity and represent the same thing, apparently.

File: pymc_marketing/mmm/incrementality.py [hunk 15]
```diff
@@ -0,0 +1,1170 @@
...
+        if frequency == "all_time":
+            return [(start, end)]
+
+        if frequency == "original":
+            # One tuple per date in the data's native frequency
+            dates = pd.date_range(
+                start,
+                end,
+                freq=pd.infer_freq(self.idata.fit_data.date.values),
+            )
```
[the remark this answers is not in the document]
Comment (isofer) [reply pymc_marketing/mmm/incrementality.py:1119]: fixed

File: pymc_marketing/mmm/incrementality.py [hunk 16]
```diff
@@ -0,0 +1,1170 @@
...
+            return [(d, d) for d in dates]
```
[the remark this answers is not in the document]
Comment (isofer) [reply pymc_marketing/mmm/incrementality.py:1120]: a non-issue

File: pymc_marketing/mmm/incrementality.py (hunk 14 above)
Comment (isofer) [reply pymc_marketing/mmm/incrementality.py:194]: discussion was moved to https://discord.com/channels/745261709622771773/948587000464834580/1472966711488544843

**Comment by isofer** [comment]
> Question: I feel we are doing something different than several other sections in the code. Here you hardcode the response variable, and calculate from the graph. **Other places request user to add the variable in the graph and the they can ask for it**.
> 
> Not critical, but this behavior could be confusing, should we stick to one no? @isofer @juanitorduz
> 
> If this is related to MMM models in the agent, we can always hardcode there to add vars in original scale (Even after the sampling if we don't want to load heavily the sampling process) but not sure, the hardcode should be here, looks again the current intuition around budget optimizer, sensitivity, and other functions using the pytensor utils.

@cetagostini just to clarify, is the issue that we hard code "channel_contribution"? Aren't we hard coding it in many other parts of the codebase?

Also, this is not related to decision AI.
cc @juanitorduz

File: pymc_marketing/mmm/incrementality.py [hunk 17]
```diff
@@ -0,0 +1,1247 @@
...
+            window_infos=window_infos,
+            max_window=max_window,
+            baseline_array=baseline_array,
+            counterfactual_spend_factor=counterfactual_spend_factor,
+            include_carryover=include_carryover,
+            l_max=l_max,
+            freq_offset=freq_offset,
+            extra_shape=extra_shape,
+            dtype=data_shared.dtype,
+        )
```
[the remark this answers is not in the document]
Comment (isofer) [reply pymc_marketing/mmm/incrementality.py:431]: fixed

File: tests/mmm/test_incrementality.py [hunk 18]
```diff
@@ -0,0 +1,695 @@
...
+            freq_offset=freq_offset,
+            extra_shape=extra_shape,
+            dtype="float64",
+        )
+
+        # With factor=0, some values should be zero (the target period dates)
+        assert (cf_array_zero == 0).any()
+        # With factor=1, arrays should be identical to each other
+        # (no change from baseline) within padded windows
+        np.testing.assert_allclose(cf_array_one, cf_array_one)
```
[the remark this answers is not in the document]
Comment (isofer) [reply tests/mmm/test_incrementality.py]: fixed

File: tests/mmm/test_incrementality.py [hunk 19]
```diff
@@ -0,0 +1,705 @@
...
+
+    Uses the standard PyMC evaluation path (completely independent from
+    extract_response_distribution + vectorize_graph) as an oracle.
+    """
+    var_name = (
+        "channel_contribution_original_scale"
+        if original_scale
+        else "channel_contribution"
+    )
+    model = cm(mmm.model)
```
Comment (ricardoV94) [inline tests/mmm/test_incrementality.py]: you can do `model = mmm.model.copy()` to avoid the import and cryptic alias

File: tests/mmm/test_incrementality.py [hunk 20]
```diff
@@ -0,0 +1,705 @@
...
+import pymc as pm
+import pytest
+import xarray as xr
+from pydantic import ValidationError
+from pymc.model.fgraph import clone_model as cm
+
+from pymc_marketing.mmm.incrementality import Incrementality
+
+# Matches simple_fitted_mmm's GeometricAdstock(l_max=4)
+_SIMPLE_L_MAX = 4
```
Comment (ricardoV94) [inline tests/mmm/test_incrementality.py]: This is used in a single place, just inline it there?

File: tests/mmm/test_incrementality.py [hunk 21]
```diff
@@ -0,0 +1,705 @@
...
+    l_max = mmm.adstock.l_max
+    inferred_freq = pd.infer_freq(dates)
+
+    # Evaluate baseline once (reused for all periods), always in original scale
+    baseline_contrib = evaluate_channel_contribution(
+        mmm, actual_data, original_scale=True
+    )
+
+    period_results = []
+    for t0, t1 in periods:
```
Comment (ricardoV94) [inline tests/mmm/test_incrementality.py:120]: This seems pretty inefficient and complex enough one almost wants a test for it. I have no immediate better proposal, but how long does `test_incrementality.py` take to run (and as a proportion of the whole CI)?

File: pymc_marketing/mmm/incrementality.py [hunk 22]
```diff
@@ -0,0 +1,1250 @@
...
+        eval_dtype = "float64"
+        batched_input = pt.tensor(
+            name="channel_data_batched",
+            dtype=eval_dtype,
+            shape=(None, *data_shared.type.shape),
+        )
+        replace_dict: dict = {data_shared: batched_input}
+        func_inputs: list = [batched_input]
+
+        # When time_varying_media (or time_varying_intercept) is enabled,
```
Comment (ricardoV94) [inline pymc_marketing/mmm/incrementality.py:364]: Opened #2330 to track this decoupling work as a follow-up.

File: pymc_marketing/mmm/incrementality.py [hunk 23]
```diff
@@ -0,0 +1,1250 @@
...
+
+            # Clip start to the requested range (needed when the user
+            # passes a start date inside a period)
+            period_start = max(period_start, start)
+
+            period_ranges.append((period_start, period_end))
+
+        return period_ranges
+
+    def _get_subsampled_idata(
```
Comment (ricardoV94) [inline pymc_marketing/mmm/incrementality.py]: This doesn't seem specific to incrementality, so maybe should live somewhere else? Or just use arviz utility

File: tests/mmm/test_incrementality.py (hunk 21 above)
Comment (isofer) [reply tests/mmm/test_incrementality.py:120]: The full `test_incrementality.py` suite takes ~94 seconds locally.

File: pymc_marketing/mmm/incrementality.py (hunk 22 above)
Comment (isofer) [reply pymc_marketing/mmm/incrementality.py:364]: Agreed — I'll follow the _set_predictors_for_optimization pattern and create a hook method on MMM for this. Prefer to do it in a follow-up to keep this PR focused. If that works for you, I'll open a ticket to track it.

File: tests/mmm/test_incrementality.py (hunk 21 above)
Comment (isofer) [reply tests/mmm/test_incrementality.py:120]: Also, this is the simplest way I found to implement it. I wanted to have an end-to-end solution that is relatively easy to read and understand, and while I don't think it's super trivial, I don't think it's complex either. It's not optimize for speed on purpose - it's a tradeoff for readability.

File: tests/mmm/test_incrementality.py (hunk 19 above)
Comment (isofer) [reply tests/mmm/test_incrementality.py]: done

File: tests/mmm/test_incrementality.py (hunk 20 above)
Comment (isofer) [reply tests/mmm/test_incrementality.py]: done

File: pymc_marketing/mmm/incrementality.py (hunk 23 above)
Comment (isofer) [reply pymc_marketing/mmm/incrementality.py]: done

File: pyproject.toml [hunk 24]
```diff
@@ -32,7 +32,7 @@ dependencies = [
     "pandas",
     "pydantic>=2.1.0",
     # NOTE: Used as minimum pymc version with test.yml `OLDEST_PYMC_VERSION`
-    "pymc>=5.27.1",
+    "pymc==5.27.1",
```
[the remark this answers is not in the document]
Comment (cetagostini) [reply pyproject.toml]: Let's avoid, givent the fact we are moving to PyMC.dims

[the remark this answers is not in the document]
Comment (isofer) [reply pyproject.toml]: @cetagostini some tests in pymc-marketing has started to break due to package dependency so I was testing a fix.

File: tests/mmm/test_mmm.py [hunk 25]
```diff
@@ -1401,10 +1401,18 @@ def test_plot_new_spend_contributions_original_scale(mmm_fitted) -> None:
 
 
 @pytest.fixture(scope="module")
-def mmm_with_prior(mmm) -> MMM:
+def mmm_with_prior() -> MMM:
     n_chains = 1
     n_samples = 100
 
+    mmm = MMM(
```
Comment (isofer) [inline tests/mmm/test_mmm.py:1408]: the fixture was overwriting mmm.idata which cause some test [failures](https://github.com/pymc-labs/pymc-marketing/actions/runs/22280820724/job/64451080148?pr=2290) when test ran in specific order: 
minimum code to reproduce localy: 
```
pytest tests/mmm/test_mmm.py::TestMMMHelperMethods::test_has_new_scaling tests/mmm/test_mmm.py::test_plot_new_spend_contributions_prior_select_channels tests/mmm/test_mmm.py::TestMMMHelperMethods::test_format_recovered_transformation_parameters_different_quantiles -v
```

Comment (juanitorduz) [reply tests/mmm/test_mmm.py:1408]: amazing!

File: .github/workflows/test.yml [hunk 26]
```diff
@@ -22,7 +22,7 @@ concurrency:
 
 env:
   # The lower bound from the pyproject.toml file
-  OLDEST_PYMC_VERSION: "$(grep -E 'pymc *>' pyproject.toml | sed -n 's/.*>=\\([0-9]*\\.[0-9]*\\.[0-9]*\\).*/\\1/p')"
+  OLDEST_PYMC_VERSION: "$(grep -E 'pymc *[>=]' pyproject.toml | sed -n 's/.*[>=]=\\([0-9]*\\.[0-9]*\\.[0-9]*\\).*/\\1/p')"
```
[the remark this answers is not in the document]
Comment (juanitorduz) [reply .github/workflows/test.yml:25]: @isofer is this change necessary?

File: pymc_marketing/data/idata/utils.py [hunk 27]
```diff
@@ -338,3 +342,66 @@ def aggregate_idata_dims(
...
+
+    rng = np.random.default_rng(random_state)
+    flat_indices = rng.choice(total_samples, size=num_samples, replace=False)
+
+    stacked = posterior.stack(sample=("chain", "draw"))
+    selected = stacked.isel(sample=flat_indices)
+    posterior_ds = (
+        selected.drop_vars(["chain", "draw", "sample"])
+        .rename({"sample": "draw"})
+        .expand_dims("chain")
```
[the remark this answers is not in the document]
Comment (juanitorduz) [reply pymc_marketing/data/idata/utils.py:398]: Maybe keep the type hints more strict?

File: pymc_marketing/mmm/incrementality.py [hunk 28]
```diff
@@ -0,0 +1,1194 @@
...
+        )
+        n_chains = posterior_sub.sizes["chain"]
+        n_draws = posterior_sub.sizes["draw"]
+
+        # Extract response distribution (batched over samples)
+        response_graph = extract_response_distribution(
+            pymc_model=self.model.model,
+            idata=az.InferenceData(posterior=posterior_sub),
+            response_variable="channel_contribution",
+        )
```
[the remark this answers is not in the document]
Comment (juanitorduz) [reply pymc_marketing/mmm/incrementality.py:333]: add_original_scale_contribution_variable just add the original scale , "channel_contribution" should always exist

File: pymc_marketing/mmm/incrementality.py (hunk 22 above)
Comment (juanitorduz) [reply pymc_marketing/mmm/incrementality.py:364]: was this resolved :) ?

File: tests/mmm/conftest.py [hunk 29]
```diff
@@ -34,38 +35,74 @@
...
+
+    Uses month-start frequency where months have variable lengths (28-31 days).
+    This exposes the _convert_frequency_to_timedelta bug that approximates
+    months as fixed 30-day periods.
+
+    Returns dict with:
+    - X: DataFrame with date and 2 channels (18 monthly periods)
+    - y: Series with target values
+    """
+    return _make_mmm_data(periods=18, freq="MS", n_channels=3, seed=99)
```
[the remark this answers is not in the document]
Comment (juanitorduz) [reply tests/mmm/conftest.py:105]: @isofer this one is easy top fix :)

**Review by juanitorduz — CHANGES_REQUESTED** [review]
@isofer I think this one is almost ready to go! I added minor comments and I thinmk we are good to go so that we work on iteratons :)

File: pymc_marketing/mmm/incrementality.py (hunk 22 above)
Comment (isofer) [reply pymc_marketing/mmm/incrementality.py:364]: No. I am suggesting not to resolve it now, but instead open a ticket to resolve it soon. This PR has been open for a week, and there two PRs that are building on it. So I would like to merge it ASAP.

File: tests/mmm/conftest.py (hunk 29 above)
[the remark this answers is not in the document]
Comment (isofer) [reply tests/mmm/conftest.py:105]: done

File: pymc_marketing/data/idata/utils.py [hunk 30]
```diff
@@ -338,3 +342,58 @@ def aggregate_idata_dims(
...
+    flat_indices = rng.choice(total_samples, size=num_samples, replace=False)
+
+    stacked = dataset.stack(sample=("chain", "draw"))
+    selected = stacked.isel(sample=flat_indices)
+
+    return (
+        selected.drop_vars(["chain", "draw", "sample"])
+        .rename({"sample": "draw"})
+        .expand_dims("chain")
+    )
```
[the remark this answers is not in the document]
Comment (isofer) [reply pymc_marketing/data/idata/utils.py:399]: it's not. you don't know what you are talking about.

File: pymc_marketing/data/idata/utils.py [hunk 31]
```diff
@@ -338,3 +342,58 @@ def aggregate_idata_dims(
...
+    >>> sub.sizes["draw"]
+    100
+    """
+    if num_samples is None:
+        return dataset
+
+    total_samples = dataset.sizes["chain"] * dataset.sizes["draw"]
+
+    if num_samples >= total_samples:
+        return dataset
```
Comment (ricardoV94) [inline pymc_marketing/data/idata/utils.py:387]: It's a bit odd that in this case you don't get the flattening of chains. You could instead say: `selected = stacked` and place the existing `rng->stacked` parts in an `else branch`

File: pymc_marketing/mmm/incrementality.py (hunk 22 above)
Comment (ricardoV94) [reply pymc_marketing/mmm/incrementality.py:364]: I don't want to block on this, but keep in mind that I'll have to change this code immediately for #2204, and it doesn't make sense for to do it twice (after this PR and after the promised refactor). How much work is it to do it now? 

The follow up PRs that depend on this, don't care about this function, just need to use it?

**Review by ricardoV94 — COMMENTED** [review]
PR looks okay. I left a comment and a follow-up on the existing discussion.

Hesitant to delay the refactor, but won't block on it if you really don't want to do it now.

**Review by ricardoV94 — APPROVED** [review]
Approving so it's in your hands @isofer

File: tests/mmm/test_incrementality.py (hunk 21 above)
Comment (ricardoV94) [reply tests/mmm/test_incrementality.py:120]: > The full test_incrementality.py suite takes ~94 seconds locally.

Is that after caching? I'm running the whole CI on my machine for the refactor into dims PR and that's 6m20, so an extra 1m30 is not nothing. 

If after caching it's faster I'm less concerned.

(Obviously we may have very different machines)

Comment (ricardoV94) [reply tests/mmm/test_incrementality.py:120]: image

Comment (isofer) [reply tests/mmm/test_incrementality.py:120]: it's 90s without cache and 30 with cache. 
For comparison I ran test_budget_optimizer.py on my machine and it took 11.3s on my machine and 15s on yours.

**Review by juanitorduz — APPROVED** [review]

File: pymc_marketing/mmm/incrementality.py (hunk 22 above)
Comment (isofer) [reply pymc_marketing/mmm/incrementality.py:364]: @ricardoV94 
I already have the next PR (#2300) ready and it heavily modifies this code so I really prefer to refactor after the merge.
I don't know how much work it will take to do the promised refactor but I'm guessing it will be non-trivial amount of time, especially since we don't have a clear design in mind so that would need to be figured out first.
In any case the code will need to be changed multiple times: (1) for #2204, (2) for the promised refactor, and (3) for #2300. So I suggest to do it in the following order to minimize conflicts:
1. Merge this one (which I'll do since I got two approvals).
2. Merge the #2300, which is waiting for this PR.
3. Merge #2204 which will change this code again.
4. I will then start working on the promised refactor.
How does it sound

Comment (isofer) [reply pymc_marketing/mmm/incrementality.py:364]: Opened #2330 to track this decoupling work as a follow-up.

**Comment by ricardoV94** [comment]
I think this PR accidentally re-reverted https://github.com/pymc-labs/pymc-marketing/pull/2324
